ENH: Integrate FastBilateralImageFilter from remote module

dzenanz opened the pull request:
"New" files are taken directly from https://github.com/InsightSoftwareConsortium/ITKFastBilateral at commit cba51ae59ae841405298784bc49b6f51a4adab03 (v1.0.1).

Modified files are merger of corresponding content from the remote module.

This is the first remote module to be integrated as part of #3371.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5

dzenanz committed:
DOC: Improve documentation of FastBilateralImageFilter

dzenanz committed:
ENH: Remove remote module option and add migration guide entry

jhlegarreta commented:
Unless all style and compiler changes that have been contributed in the past 2 to 3 months by Hans and Niels are applied to this code, we will be introducing code having a discouraged style.

dzenanz commented:
There are multiple remote modules in the same boat. Should we integrate them, and then make code style changes pass on the entire source, including the newly added code? Or require those style changes be made before integrating? The second option will probably require more effort. @N-Dekker and @hjmjohnson might want to pitch in.

N-Dekker commented:
Is the FastBilateralImageFilter still relevant nowadays? I see, there was a publication of the filter by Jordan Woehr on 2009-08-28, [A Fast Approximation to the Bilateral Filter for ITK](https://insight-journal.org/browse/publication/692). The remote module was originally by Pablo (@phcerdan) https://github.com/InsightSoftwareConsortium/ITKFastBilateral/commit/451baba6d77f85c0a4ffccfbbde1adc0acd9721b

In general I would suggest people to double-check if a remote module is still relevant, before integrating them into the core ITK Modules. Otherwise ITK will just grow and grow, increasing the maintenance burden and increasing build times.

N-Dekker commented on Modules/Filtering/ImageFeature/include/itkFastBilateralImageFilter.h:
```diff
@@ -0,0 +1,204 @@
+ * \todo Support for color images
+ * \todo Support for vector images
+ */
+
+template <class TInputImage, class TOutputImage>
```
ITK started using `typename` for template parameters back in 2013, especially commit 86bbb9d092d21bb309b0260c0b4082d9af966b67 by Hans (@hjmjohnson)

N-Dekker commented on Modules/Filtering/ImageFeature/include/itkFastBilateralImageFilter.h:
```diff
@@ -0,0 +1,204 @@
+  using InterpolatorType = LinearInterpolateImageFunction<GridType, float>;
+  using InterpolatedIndexType = typename InterpolatorType::ContinuousIndexType;
+
+  double               m_RangeSigma;
+  DomainSigmaArrayType m_DomainSigma;
```
Nowadays we would use data member initializers:

    double               m_RangeSigma{ 50.0 };
    DomainSigmaArrayType m_DomainSigma{ DomainSigmaArrayType::Filled(40.0) };

N-Dekker commented on Modules/Filtering/ImageFeature/include/itkFastBilateralImageFilter.h:
```diff
@@ -0,0 +1,204 @@
+    m_RangeSigma = 50.0;
+  }
+
+  virtual ~FastBilateralImageFilter() {}
```
Nowadays:

    ~FastBilateralImageFilter() override = default;

N-Dekker commented on Modules/Filtering/ImageFeature/test/CMakeLists.txt:
```diff
@@ -13,6 +13,9 @@ set(ITKImageFeatureTests
     itkBilateralImageFilterTest.cxx
     itkBilateralImageFilterTest2.cxx
     itkBilateralImageFilterTest3.cxx
+    itkFastBilateralImageFilterTest.cxx
+    itkFastBilateralImageFilterTest2.cxx
+    itkFastBilateralImageFilterTest3.cxx
```
I'm not a fan of test names like "Test2", "Test3", etc. Hope we can eventually move towards more GTest style test, and use more meaningful names for tests.

N-Dekker commented on Modules/Filtering/ImageFeature/test/itkFastBilateralImageFilterTest3.cxx:
```diff
@@ -0,0 +1,92 @@
+    return -1;
+  }
+
+  // Generate test image
+  itk::ImageFileWriter<ImageType>::Pointer writer;
+  writer = itk::ImageFileWriter<ImageType>::New();
+  writer->SetInput(filter3->GetOutput());
+  writer->SetFileName(av[2]);
+  writer->Update();
```
Nowadays just one line of code: 

    itk::WriteImage(filter3->GetOutput(), av[2]);

N-Dekker commented on Modules/Filtering/ImageFeature/test/itkFastBilateralImageFilterTest2.cxx:
```diff
@@ -0,0 +1,93 @@
+    return -2;
+  }
+
+  // Generate test image
+  itk::ImageFileWriter<ImageType>::Pointer writer;
+  writer = itk::ImageFileWriter<ImageType>::New();
+  writer->SetInput(filter->GetOutput());
+  writer->SetFileName(av[2]);
+  writer->Update();
```
Nowadays just one line of code: 

    itk::WriteImage(filter3->GetOutput(), av[2]);

N-Dekker commented on Modules/Filtering/ImageFeature/test/itkFastBilateralImageFilterTest.cxx:
```diff
@@ -0,0 +1,100 @@
+  filter->AddObserver(itk::ProgressEvent(), showProgress);
+  filter->SetInput(image);
+
+  using WriterType = itk::ImageFileWriter<ImageType>;
+  WriterType::Pointer writer = WriterType::New();
+  writer->SetFileName(outputImageFileName);
+  writer->SetInput(filter->GetOutput());
+  writer->SetUseCompression(true);
+
+  ITK_TRY_EXPECT_NO_EXCEPTION(writer->Update());
```
Nowadays just one line of code: 

    itk::WriteImage(filter->GetOutput(), outputImageFileName, true);

Is `ITK_TRY_EXPECT_NO_EXCEPTION still necessary? With GTest, a unit test that throw an exception (uncaught) is automatically reported as test failure.

dzenanz commented:
Niels, it would be good if you tackled these (and possibly other changes) yourself. Push changes either to my branch from which this PR is open, or in the current remote module (I would pick up the changes from there later and integrate them in this PR).

N-Dekker commented on Modules/Filtering/ImageFeature/include/itkFastBilateralImageFilter.hxx:
```diff
@@ -0,0 +1,311 @@
+  // We now have an empty grid. Two will be needed total,
+  // one will be a container for the down-sampled data and one
+  // will remember how many pixels were placed into each bin
+  gridImage->Allocate();
+  // Init values of image to 0
+  gridImage->FillBuffer(0.0);
```
Nowadays, no need to call `Allocate(); FillBuffer(0.0)`, just call:

    gridImage->AllocateInitialized();

N-Dekker commented on Modules/Filtering/ImageFeature/include/itkFastBilateralImageFilter.hxx:
```diff
@@ -0,0 +1,311 @@
+    InputPixelType intensityDelta = static_cast<InputPixelType>(intensityMax - intensityMin);
+    gridSize[ImageDimension] = static_cast<GridSizeValueType>((intensityDelta / m_RangeSigma) + 1 + 2 * padding);
+
+    for (size_t i = 0; i < ImageDimension + 1; ++i)
+    {
+      gridStartPos[i] = 0;
+    }
+
+    GridRegionType region;
+    region.SetSize(gridSize);
+    region.SetIndex(gridStartPos);
+    gridImage->SetRegions(region);
```
I would remove both local variables `gridStartPos` and `region`. `gridStartPos` just contains zero's. The region size of the gridImage can be specified directly, without local `region` variable, by:

    gridImage->SetRegions(gridSize);

N-Dekker commented:
> Niels, it would be good if you tackled these (and possibly other changes) yourself.

@dzenanz  Sorry, I'm just doing a little bit of reviewing of this PR now. By the way, none of the comments I made so far look like showstoppers to me. They could also be addressed after merging. Or you could directly address them yourself.

blowekamp commented:
Could we consider an alternative name to "FastBilateralImageFilter". Anything more descriptive of the name? If another version is created that is even faster, how would it be distinguished?

dzenanz commented:
I would prefer not to rename it during transition. Having the same class name will make it easier for people to find it. We could rename it in the future, either with next major version of ITK (e.g. v7), or when another implementation appears.

jhlegarreta commented:
> There are multiple remote modules in the same boat. Should we integrate them, and then make code style changes pass on the entire source, including the newly added code? Or require those style changes be made before integrating? The second option will probably require more effort. @N-Dekker and @hjmjohnson might want to pitch in.

IMO, all such changes should be applied to the remotes prior to integrating them, even when integrating them is not yet on the roadmap. We tend to forget about examples, modules not turned on by default and remotes. The later may be more difficult, but there have been multiple efforts over time to try to apply sweeping changes trying to avoid the burden it implies doing it one by one/separately/manually on each local clone, e.g.

https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/Maintenance/ApplyScriptToRemotes.sh

From issue #4748:
"Also, there is this tool Matt has used to update the remote module CI config to ITK 5.4:
https://github.com/asottile/all-repos

Ongoing work:
https://github.com/thewtex/ITK/commit/9e9d5131dcde8f46505c838218df68ce24701cd0
"

dzenanz committed:
ENH: Integrate FastBilateralImageFilter from remote module (change)

"New" files are taken directly from
https://github.com/InsightSoftwareConsortium/ITKFastBilateral
at commit cba51ae59ae841405298784bc49b6f51a4adab03 (v1.0.1).

Modified files are merger of corresponding content from the remote module.

dzenanz commented:
Many changes were applied to the remote. I updated the PR, using the newest code from the remote. If we want to do some code style updates or other refactoring, that can be done either prior (in the remote) or after the merge (here, in main repo). Please re-review.

hjmjohnson commented:
This PR is superseded by #6060 (hierarchical category repos). FastBilateral will be consolidated into **ITKRemoteFiltering** rather than integrated into ITK core.

Reviewing the concerns raised here against the current state of [ITKFastBilateral](https://github.com/InsightSoftwareConsortium/ITKFastBilateral):

| Concern | Status |
|---|---|
| Style/compiler changes before integration (@jhlegarreta) | **Done** — clang-format synced (#11), `itkStaticConstMacro` → `constexpr` (#9), `itkOverrideGetNameOfClassMacro` (#8), `using` declarations already in place |
| Still relevant? (@N-Dekker) | **Yes** — confirmed by @dzenanz (used in production code) |
| Rename "FastBilateral" to something more descriptive (@blowekamp) | **Deferred** — @dzenanz preferred keeping the name for discoverability; rename can happen in a future major version |
| Update to modern standards/conventions (@blowekamp) | **Done** — CI updated (#13, #12), Python packaging modernized (#5) |
| Remaining cosmetic: comment text says "typedefs" (lines 78, 94, 105) | Minor — can be fixed during consolidation |

The consolidation approach (#6060) addresses @jhlegarreta's broader concern about applying sweeping changes: all modules in a category repo share one CI, one `.clang-format`, and one set of style enforcement — so future style updates are a single PR instead of 60.

hjmjohnson commented:
@dzenanz I've redone this PR using the v3 whitelist-based ingestion strategy now applied to all remote-module consolidations: see #6159. The new PR lands FastBilateral as its own `Modules/Filtering/FastBilateral/` (matching #6135 / #6137 for GenericLabelInterpolator / MGHIO) instead of merging the files into `Modules/Filtering/ImageFeature/`, and preserves the upstream contributors as `git blame` authors via a `filter-repo` Mode A merge.

Once #6159 is reviewed and merged, this PR can be closed (or you can close it now if you'd prefer the new shape).
